Source/Dafny/Translator.cs

```diff
@@ -94,6 +96,229 @@ public string FreshNumericId(string prefix = "")
+        return String.Format("({0})", sub);
+      } else {
+        // other expressions are treated as opaque
+        // we offer limited support to create temp variables that represent their values
+        var sub = this.OpacifySubExpr(expr);
+        if (!this.opaqueExprs.ContainsKey(sub)) {
+          var temp = this.AllocateFreshTemp();
+          this.opaqueExprs.Add(sub, temp);
+        }
+        return this.opaqueExprs[sub];
+      }
+    }
+
+    string EncodeEqualityPoly(Expression e0, Expression e1) {
+      var left = this.EncodeSubExpr(e0);
+      var right = this.EncodeSubExpr(e1);
+      var poly = String.Format("{0} - {1}", left, right);
+      return poly;
+    }
+
+    void EncodeGeneratorPoly(Expression expr) {
+      if (expr is BinaryExpr) {
+        var binExpr = expr as BinaryExpr;
+        if (binExpr.Op != BinaryExpr.Opcode.Eq) {
+          Console.WriteLine("Error: Grobner generator expression should be an Equality or IsModEquivalent");
+          throw new cce.UnreachableException();
+        }
+        polys.Add(this.EncodeEqualityPoly(binExpr.E0, binExpr.E1));
+      } else if (expr is ApplySuffix) {
+        var appplyExpr = expr as ApplySuffix;
+        if (!(appplyExpr.Lhs is NameSegment) || !(appplyExpr.Lhs as NameSegment).Name.Equals("IsModEquivalent")) {
+          Console.WriteLine("Error: Grobner generator expression should be an Equality or IsModEquivalent");
+          throw new cce.UnreachableException();
+        }
+        var left = this.EncodeSubExpr(appplyExpr.Args[0]);
+        var right = this.EncodeSubExpr(appplyExpr.Args[1]);
+        var mod = this.EncodeSubExpr(appplyExpr.Args[2]);
+        var temp = this.AllocateFreshTemp();
+        var poly = String.Format("{0} - {1} - {2} * {3}", left, right, mod, temp);
+        polys.Add(poly);
+      } else {
+        Console.WriteLine("unhandled generator expr: {0}", expr);
+        throw new cce.UnreachableException();
+      }
+    }
+
+    string EncodeQueryPoly(Expression expr) {
+      if (expr is BinaryExpr) {
+        var binExpr = expr as BinaryExpr;
+        if (binExpr.Op != BinaryExpr.Opcode.Eq) {
+          Console.WriteLine("Error: Grobner query expression should be an Equality or IsModEquivalent");
+          throw new cce.UnreachableException();
+        }
+        return this.EncodeEqualityPoly(binExpr.E0, binExpr.E1);
+      } else if (expr is ApplySuffix) {
+        var appplyExpr = expr as ApplySuffix;
+        if (!(appplyExpr.Lhs is NameSegment) || !(appplyExpr.Lhs as NameSegment).Name.Equals("IsModEquivalent")) {
+          Console.WriteLine("Error: Grobner query expression should be an Equality or IsModEquivalent");
+          throw new cce.UnreachableException();
+        }
+        var mod = this.EncodeSubExpr(appplyExpr.Args[2]);
+        // add the mod to the generators
+        polys.Add(mod); 
+        return this.EncodeEqualityPoly(appplyExpr.Args[0], appplyExpr.Args[1]);
```

Comment:
Why aren't `Args[0]` and `Args[1]` encoded here?  They were encoded in `EncodeGeneratorPoly`.

Reply:
Oh in this case the `Args[2]` needs to go into the ideal. The rest of the polys in the generators are effectively all 0, except this one. Then if the difference `Args[0] - Args[1]` can be written in terms of a linear combination of the generators, it must be a multiple of `Args[2]`, proving the congruence.